### PR TITLE
feat(extension): surface engine merge-queue state via the MQ status comment payload

### PR DESCRIPTION
**Pull request by jd**
Engine PR Mergifyio/monorepo#32922 embeds a JSON merge-queue state payload
in the MQ status comment (it can't live in the check-run summary because
GitHub trims that HTML, and the check-runs API needs auth the extension
doesn't have). The marker block lands in the conversation DOM — but
GitHub's markdown sanitizer strips HTML comments from the *rendered*
timeline. The raw markdown source does survive in the inline edit-form
`<textarea class="CommentBox-input">` GitHub hydrates for inline editing,
and is also fetchable via `/issue_comments/<id>/edit_form` for comments
whose textarea isn't materialized — same pattern stacks.js already uses
for stack comments.

## What ships

**src/mqPayload.js (new)** — finds the latest Mergify status comment
across `.TimelineItem` nodes (reverse walk to skip Mergify-authored
non-status events that come after), parses the marker block, exposes
`attach(callback) / detach() / getCurrent()`. Live updates via a
MutationObserver on `document.body` (debounced 800ms, short-circuits
once in payload mode since the engine edits the same comment in place)
plus a 15s periodic poll as the safety net (inline `.value` mutations
on a `<textarea>` don't fire MutationObserver). Falls back to the
existing `/checks-page` scrape when no Mergify status comment exists
(`status_comments: none`, older on-prem engines, pre-comment window),
and upgrades to payload mode automatically as soon as one appears.
Two generation counters guard against stale in-flight emits across
detach or upgrade transitions.

**src/queue.js** — new "rich" row rendered from the payload: yellow
CI-in-progress Checking pill (amber Frozen, red Bisecting, green
Merged, red Dequeued, grey Waiting), live ETA countdown, Mergify-
branded state row, Batch PR button for the speculative-check PR,
and a compact "Blocking the merge (N):" list of unmet conditions
(met conditions skipped entirely; "any of" / "not" groups collapse
to a single inline row so the disjunction stays visible). The old
single-row layout becomes the "legacy" shape when no payload is
available. `buildMergifyRow(variant)` / `updateMergifyRow(row)`
keep PR #494's variant-based identity (`MERGE_BOX_ROW_ATTR`) and
dispatch internally to rich-vs-legacy based on `mqPayload.getCurrent()`.

**src/mergify.js** — subscribes via `mqPayload.attach`, runs alongside
PR #494's `fetchQueueStateIfNeeded` + `scheduleQueueStatePoll` so the
legacy poll path keeps driving updates when no payload exists. Side
peek + bottom mergebox stay in sync via `updateAllMergifyRows`.
`resetForNavigation` tears down mqPayload alongside the existing
queue/stack reset.

## Test surface

224 unit tests across 4 suites (Jest + jsdom). Coverage includes:
parser/regex edge cases, the textarea finder with inline + edit_form
fallback, attach/detach lifecycle including race-condition regressions
(zombie poll after upgrade, stale emit after detach), conditions block
rendering for met/unmet/any-of/all-of, ETA formatter boundaries, rich
row rendering for all six states, ETA ticker that fans out to all
mounted rows, multi-variant identity.

## Verified on

- #33609 (queued/checking with ETA + Batch PR)
- #32178 (in queue, edit_form fallback path exercised)
- #32275 (no MQ status comment → legacy fallback)
- Side peek opened from "View status" button shows the same row as the
  bottom mergebox

Co-Authored-By: Claude Opus 4.7 (1M context) <noreply@anthropic.com>

### CODE REVIEW

**Comment by jd** [comment]
### Revision history

| # | Type | Changes | Reason | Date |
|---|------|---------|--------|------|
| 1 | initial | `43af264` |  | 2026-06-05 06:25 UTC |
| 2 | content | [`43af264 → 7c38577`](https://github.com/Mergifyio/browser-extension/compare/43af2645a4673ac581b929c65299b70c6ecab3a8...5f783618c95d9658f79717c41b900b7fdca4d84a) ([raw](https://github.com/Mergifyio/browser-extension/compare/43af2645a4673ac581b929c65299b70c6ecab3a8...7c38577d0e8d3d5564a55548f42a6420b6a7f379)) |  | 2026-06-05 06:25 UTC |
| 3 | content | [`7c38577 → 551c43d`](https://github.com/Mergifyio/browser-extension/compare/7c38577d0e8d3d5564a55548f42a6420b6a7f379...c2b951f5318c5a8e767ab575335c919ca502a1bd) ([raw](https://github.com/Mergifyio/browser-extension/compare/7c38577d0e8d3d5564a55548f42a6420b6a7f379...551c43d7c42a31155e79c0e6222ad6f905ec31e8)) | format queued duration as Hh Mm / Dd Hh past 1h / 1d so long queue waits stay legible (e.g. 822m → 13h 42m) | 2026-06-05 07:19 UTC |

File: src/queue.js [hunk 1]
```diff
@@ -1,13 +1,27 @@
 import { MergifyCache } from "./cache.js";
 import { debug } from "./debug.js";
-import { getPullRequestData, isGitHubPullRequestPage } from "./dom.js";
+import { getPullRequestData } from "./dom.js";
```
[the remark this answers is not in the document]
Comment (jd) [reply src/queue.js]: Good catch — confirmed runtime ReferenceError. `isGitHubPullRequestPage` got dropped from the imports during the rebase onto PR #494 (my pre-reconcile queue.js didn't need it because the fallback poll was in mqPayload.js; when I brought `scheduleQueueStatePoll` back from upstream I missed restoring the import). Added it back to the import block, and added a regression test that runs `scheduleQueueStatePoll()` under fake timers and advances past `QUEUE_POLL_INTERVAL` — would have caught this.

File: src/queue.js [hunk 2]
```diff
@@ -574,48 +585,468 @@ function buildMergifySidebarRow() {
...
+            : buildLegacyRow(variant);
+        row.replaceWith(replacement);
+        debug("Mergify row swapped:", currentShape, "→", incomingShape);
+        return;
+    }
+
+    if (incomingShape === "rich") {
+        const next = buildRichRow(payload, variant);
+        row.replaceChildren(...next.childNodes);
+        _etaPayload = payload;
```
[the remark this answers is not in the document]
Comment (jd) [reply src/queue.js:632]: Technically the spread eagerly iterates the NodeList before `replaceChildren` starts moving nodes (per the JS spec: `f(...iter)` consumes `iter` to completion before calling `f`), so this is safe in practice. But you're right that the pattern is fragile to read — switched to `Array.from(next.childNodes)` to make the snapshot explicit so a future reader (or static analyzer) doesn't need to reason about iteration-during-mutation.

**Review by JulianMaurin — DISMISSED** [review]

**Review by sileht — APPROVED** [review]

**Review by remyduthu — APPROVED** [review]
